src/outputs/histogram.cpp

```diff
@@ -0,0 +1,554 @@
+              x_bin = static_cast<int>((x_val - x_edge_min) / x_edge_dbin);
+            } else if (x_edges_type == EdgeType::Log) {
+              x_bin = static_cast<int>((Kokkos::log10(x_val) - x_edge_min) / x_edge_dbin);
+              // otherwise search
+            } else {
+              x_bin = upper_bound(x_edges, x_val) - 1;
+            }
+          }
+          PARTHENON_DEBUG_REQUIRE(x_bin >= 0, "Bin not found");
+
+          // needs to be zero as for the 1D histogram we need 0 as first index of the 2D
+          // result array
+          int y_bin = 0;
+          if (hist_ndim == 2) {
+            auto y_val = std::numeric_limits<Real>::quiet_NaN();
+            if (y_var_type == VarType::X1) {
+              y_val = coords.Xc<1>(k, j, i);
+            } else if (y_var_type == VarType::X2) {
+              y_val = coords.Xc<2>(k, j, i);
+            } else if (y_var_type == VarType::X3) {
+              y_val = coords.Xc<3>(k, j, i);
+            } else if (y_var_type == VarType::R) {
+              y_val =
+                  Kokkos::sqrt(SQR(coords.Xc<1>(k, j, i)) + SQR(coords.Xc<2>(k, j, i)) +
+                               SQR(coords.Xc<3>(k, j, i)));
+            } else {
+              y_val = y_var(b, y_var_component, k, j, i);
+            }
+
+            if (y_val < y_edges(0) || y_val > y_edges(y_edges.extent_int(0) - 1)) {
+              return;
+            }
+
+            y_bin = -1; // reset to impossible value
+            // if we're on the rightmost edge, directly set last bin
+            if (y_val == y_edges(y_edges.extent_int(0) - 1)) {
+              y_bin = y_edges.extent_int(0) - 2;
+            } else {
+              // for lin and log directly pick index
+              if (y_edges_type == EdgeType::Lin) {
+                y_bin = static_cast<int>((y_val - y_edge_min) / y_edge_dbin);
+              } else if (y_edges_type == EdgeType::Log) {
+                y_bin =
+                    static_cast<int>((Kokkos::log10(y_val) - y_edge_min) / y_edge_dbin);
+                // otherwise search
+              } else {
+                y_bin = upper_bound(y_edges, y_val) - 1;
+              }
+            }
+            PARTHENON_DEBUG_REQUIRE(y_bin >= 0, "Bin not found");
+          }
+          auto res = scatter.access();
+          const auto val_to_add = binned_var_component == -1
+                                      ? 1
+                                      : binned_var(b, binned_var_component, k, j, i);
+          auto weight = weight_by_vol ? coords.CellVolume(k, j, i) : 1.0;
+          weight *= weight_var_component == -1
+                        ? 1.0
+                        : weight_var(b, weight_var_component, k, j, i);
+          res(y_bin, x_bin) += val_to_add * weight;
+        });
+    // "reduce" results from scatter view to original view. May be a no-op depending on
+    // backend.
+    Kokkos::Experimental::contribute(result.KokkosView(), scatter);
```

Comment:
Cool! I assume this is more performant than atomic ads?

Reply:
The `ScatterView` transparently switches between using atomics and temporary arrays depending on the backend, so this is actually be a no-op if atomics were used inside the kernel.

Reply:
Neat! TIL